src/effects/deeplearning/ModelCard.cpp

```diff
@@ -0,0 +1,340 @@
+
+      rapidjson::StringBuffer sb;
+      validator.GetInvalidSchemaPointer().StringifyUriFragment(sb);
+      message += "violation found in URI: " + std::string(sb.GetString()) + "\n";
+      message += "the following schema field was violated: " 
+                  + std::string(validator.GetInvalidSchemaKeyword()) + "\n";
+      sb.Clear();
+
+      rapidjson::Writer<rapidjson::StringBuffer> writer(sb);
+      doc->Accept(writer);
+      message += "invalid document: \n\t" 
+                  + std::string(sb.GetString()) + "\n";
+
+      sb.Clear();
+      rapidjson::Writer<rapidjson::StringBuffer> swriter(sb);
+      schema->Accept(swriter);
+      message += "schema document: \n\t" 
+                  + std::string(sb.GetString()) + "\n";
+
+      throw InvalidModelCardDocument(message.c_str(), doc);
+   }
+}
+
+void ModelCard::SerializeToFile(const std::string &path) const
+{
+   rapidjson::StringBuffer sb;
+   rapidjson::Writer<rapidjson::StringBuffer> writer(sb);
+
+   writer.StartObject();
+   Serialize(writer);
+   writer.EndObject();
+
+   wxFile file(path, wxFile::write);
+   file.Write(wxString(sb.GetString()));
+}
+
+void ModelCard::DeserializeFromFile(const std::string &path, DocHolder schema)
+{
+   DocHolder d = parsers::ParseFile(path);
+   Deserialize(d, schema);
+}
+
+template < typename Writer >
+void ModelCard::Serialize(Writer &writer) const
+{
+   // name
+   writer.String("name");
+   writer.String(m_name.c_str());
+
+   // author
+   writer.String("author");
+   writer.String(m_author.c_str());
+   
+   // long description
+   writer.String("long_description");
+   writer.String(m_long_description.c_str());
+
+   // short description
+   writer.String("short_description");
+   writer.String(m_short_description.c_str());
+
+   // sample rate
+   writer.String("sample_rate");
+   writer.Int(m_sample_rate);
+
+   // multichannel
+   writer.String("multichannel");
+   writer.Bool(m_multichannel);
+
+   // effect type
+   writer.String("effect_type");
+   writer.String(m_effect_type.c_str());
+
+   // domain tags
+   writer.String("domain_tags");
+   writer.StartArray();
+   for (auto tag : m_domain_tags)
+      writer.String(tag.c_str());
+   writer.EndArray();
+
+   // tags
+   writer.String("tags");
+   writer.StartArray();
+   for (auto tag : m_tags)
+      writer.String(tag.c_str());
+   writer.EndArray();
+
+   // labels
+   writer.String("labels");
+   writer.StartArray();
+   for(auto label : m_labels)
+      writer.String(label.c_str());
+   writer.EndArray();
+
+   // model size
+   writer.String("model_size");
+   writer.Uint64((uint64_t)m_model_size);
+
+}
+
+void ModelCard::Deserialize(DocHolder doc, DocHolder schema)
+{
+   using namespace validators;
+   Validate(doc, schema);
+
+   // these fields are not in HF metadata but rather added later,
+   // so don't throw if they are not present
+   try
+   {
+      m_author = tryGetString("author", doc);
```

Comment:
In this case, I prefer having a no-throw overload, that provides a default value.

Like:
```
std::string tryGetString(std::string key, std::string default, Doc doc);
```

Also, passing the document as a first argument would allow defaulting the default value!